# NLP Lab A_1_b

# Understanding how to clean the dataset and Ham Spam classification using Naive Bayes

---
21BAI1056
SHREYA ROSHAN




Reading the data

In [ ]:
import pandas as pd
d=pd.read_csv('/content/spam.csv', encoding='latin-1') #READING THE DATA
d.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

Removing NaN columns

In [ ]:
pd.set_option('display.max_colwidth',1)#removing NAN columns
d=d[['v1','v2']]
print(d.head)

<bound method NDFrame.head of         v1  \
0     ham    
1     ham    
2     spam   
3     ham    
4     ham    
...   ...    
5567  spam   
5568  ham    
5569  ham    
5570  ham    
5571  ham    

                                                                                                                                                                     v2  
0     Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...                                                    
1     Ok lar... Joking wif u oni...                                                                                                                                      
2     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's        
3     U dun say so early hor... U c already then say...                                                                  

In [ ]:
print(d['v1'].value_counts())

v1
ham     4825
spam    747 
Name: count, dtype: int64


Removing punctuations

In [ ]:
import string
string
def re_punctuation(text):
  p="".join([i for i in text if i not in string.punctuation])
  return p

In [ ]:
d['a']=d['v2'].apply(lambda x:re_punctuation(x)) #removing punctuations
d.head()

v1  \
0  ham    
1  ham    
2  spam   
3  ham    
4  ham    

                                                                                                                                                            v2  \
0  Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...                                               
1  Ok lar... Joking wif u oni...                                                                                                                                 
2  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's   
3  U dun say so early hor... U c already then say...                                                                                                             
4  Nah I don't think he goes to usf, he lives around here though                                                                                                 

                                                                                                                                                       a  
0  Go until jurong point crazy Available only in bugis n great world la e buffet Cine there got amore wat                                                 
1  Ok lar Joking wif u oni                                                                                                                                
2  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive entry questionstd txt rateTCs apply 08452810075over18s  
3  U dun say so early hor U c already then say                                                                                                            
4  Nah I dont think he goes to usf he lives around here though

Generalize by conversion to lowercase

In [ ]:
d['l']=d['a'].apply(lambda x: x.lower()) #conversion to lower case
d.head()

v1  \
0  ham    
1  ham    
2  spam   
3  ham    
4  ham    

                                                                                                                                                            v2  \
0  Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...                                               
1  Ok lar... Joking wif u oni...                                                                                                                                 
2  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's   
3  U dun say so early hor... U c already then say...                                                                                                             
4  Nah I don't think he goes to usf, he lives around here though                                                                                                 

                                                                                                                                                       a  \
0  Go until jurong point crazy Available only in bugis n great world la e buffet Cine there got amore wat                                                  
1  Ok lar Joking wif u oni                                                                                                                                 
2  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive entry questionstd txt rateTCs apply 08452810075over18s   
3  U dun say so early hor U c already then say                                                                                                             
4  Nah I dont think he goes to usf he lives around here though                                                                                             

                                                                                                                                                       l  
0  go until jurong point crazy available only in bugis n great world la e buffet cine there got amore wat                                                 
1  ok lar joking wif u oni                                                                                                                                
2  free entry in 2 a wkly comp to win fa cup final tkts 21st may 2005 text fa to 87121 to receive entry questionstd txt ratetcs apply 08452810075over18s  
3  u dun say so early hor u c already then say                                                                                                            
4  nah i dont think he goes to usf he lives around here though

perform tokenization


In [ ]:
import re #regular expression matching
def tok(text):
  t=re.split('\W+',text) #White space tokenization
  return t

In [ ]:
d['token']=d['l'].apply(lambda x:tok(x))
d.head()

v1  \
0  ham    
1  ham    
2  spam   
3  ham    
4  ham    

                                                                                                                                                            v2  \
0  Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...                                               
1  Ok lar... Joking wif u oni...                                                                                                                                 
2  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's   
3  U dun say so early hor... U c already then say...                                                                                                             
4  Nah I don't think he goes to usf, he lives around here though                                                                                                 

                                                                                                                                                       a  \
0  Go until jurong point crazy Available only in bugis n great world la e buffet Cine there got amore wat                                                  
1  Ok lar Joking wif u oni                                                                                                                                 
2  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive entry questionstd txt rateTCs apply 08452810075over18s   
3  U dun say so early hor U c already then say                                                                                                             
4  Nah I dont think he goes to usf he lives around here though                                                                                             

                                                                                                                                                       l  \
0  go until jurong point crazy available only in bugis n great world la e buffet cine there got amore wat                                                  
1  ok lar joking wif u oni                                                                                                                                 
2  free entry in 2 a wkly comp to win fa cup final tkts 21st may 2005 text fa to 87121 to receive entry questionstd txt ratetcs apply 08452810075over18s   
3  u dun say so early hor u c already then say                                                                                                             
4  nah i dont think he goes to usf he lives around here though                                                                                             

                                                                                                                                                                                token  \
0  [go, until, jurong, point, crazy, available, only, in, bugis, n, great, world, la, e, buffet, cine, there, got, amore, wat]                                                          
1  [ok, lar, joking, wif, u, oni]                                                                                                                                                       
2  [free, entry, in, 2, a, wkly, comp, to, win, fa, cup, final, tkts, 21st, may, 2005, text, fa, to, 87121, to, receive, entry, questionstd, txt, ratetcs, apply, 08452810075over18s]   
3  [u, dun, say, so, early, hor, u, c, already, then, say]                                                                                                                              
4  [nah, i, dont, think, he, goes, to, usf, he, lives, around, here, though]                                                                                          

Remove stop words

In [ ]:
import nltk
nltk.download('stopwords')
sw=nltk.corpus.stopwords.words('english') #getting all english stopwords
print(sw[0:10]) #printing top stopwords

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're"]


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [ ]:
def stw(text):
  swo=[i for i in text if i not in sw]
  return swo

In [ ]:
d['swr']=d['token'].apply(lambda x:stw(x)) #removing all stop words
d.head()

v1  \
0  ham    
1  ham    
2  spam   
3  ham    
4  ham    

                                                                                                                                                            v2  \
0  Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...                                               
1  Ok lar... Joking wif u oni...                                                                                                                                 
2  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's   
3  U dun say so early hor... U c already then say...                                                                                                             
4  Nah I don't think he goes to usf, he lives around here though                                                                                                 

                                                                                                                                                       a  \
0  Go until jurong point crazy Available only in bugis n great world la e buffet Cine there got amore wat                                                  
1  Ok lar Joking wif u oni                                                                                                                                 
2  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive entry questionstd txt rateTCs apply 08452810075over18s   
3  U dun say so early hor U c already then say                                                                                                             
4  Nah I dont think he goes to usf he lives around here though                                                                                             

                                                                                                                                                       l  \
0  go until jurong point crazy available only in bugis n great world la e buffet cine there got amore wat                                                  
1  ok lar joking wif u oni                                                                                                                                 
2  free entry in 2 a wkly comp to win fa cup final tkts 21st may 2005 text fa to 87121 to receive entry questionstd txt ratetcs apply 08452810075over18s   
3  u dun say so early hor u c already then say                                                                                                             
4  nah i dont think he goes to usf he lives around here though                                                                                             

                                                                                                                                                                                token  \
0  [go, until, jurong, point, crazy, available, only, in, bugis, n, great, world, la, e, buffet, cine, there, got, amore, wat]                                                          
1  [ok, lar, joking, wif, u, oni]                                                                                                                                                       
2  [free, entry, in, 2, a, wkly, comp, to, win, fa, cup, final, tkts, 21st, may, 2005, text, fa, to, 87121, to, receive, entry, questionstd, txt, ratetcs, apply, 08452810075over18s]   
3  [u, dun, say, so, early, hor, u, c, already, then, say]                                                                                                                              
4  [nah, i, dont, think, he, goes, to, usf, he, lives, around, here, though]                                                                                          

Stemming

In [ ]:
from nltk.stem import PorterStemmer
p=PorterStemmer()
def stemming(text):
  st=[p.stem(i) for i in text]
  return st


In [ ]:
d['s']=d['swr'].apply(lambda x:stemming(x)) #bringing all words to their root form
d.head()


v1  \
0  ham    
1  ham    
2  spam   
3  ham    
4  ham    

                                                                                                                                                            v2  \
0  Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...                                               
1  Ok lar... Joking wif u oni...                                                                                                                                 
2  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's   
3  U dun say so early hor... U c already then say...                                                                                                             
4  Nah I don't think he goes to usf, he lives around here though                                                                                                 

                                                                                                                                                       a  \
0  Go until jurong point crazy Available only in bugis n great world la e buffet Cine there got amore wat                                                  
1  Ok lar Joking wif u oni                                                                                                                                 
2  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive entry questionstd txt rateTCs apply 08452810075over18s   
3  U dun say so early hor U c already then say                                                                                                             
4  Nah I dont think he goes to usf he lives around here though                                                                                             

                                                                                                                                                       l  \
0  go until jurong point crazy available only in bugis n great world la e buffet cine there got amore wat                                                  
1  ok lar joking wif u oni                                                                                                                                 
2  free entry in 2 a wkly comp to win fa cup final tkts 21st may 2005 text fa to 87121 to receive entry questionstd txt ratetcs apply 08452810075over18s   
3  u dun say so early hor u c already then say                                                                                                             
4  nah i dont think he goes to usf he lives around here though                                                                                             

                                                                                                                                                                                token  \
0  [go, until, jurong, point, crazy, available, only, in, bugis, n, great, world, la, e, buffet, cine, there, got, amore, wat]                                                          
1  [ok, lar, joking, wif, u, oni]                                                                                                                                                       
2  [free, entry, in, 2, a, wkly, comp, to, win, fa, cup, final, tkts, 21st, may, 2005, text, fa, to, 87121, to, receive, entry, questionstd, txt, ratetcs, apply, 08452810075over18s]   
3  [u, dun, say, so, early, hor, u, c, already, then, say]                                                                                                                              
4  [nah, i, dont, think, he, goes, to, usf, he, lives, around, here, though]                                                                                          

Vectorize the data

In [54]:
from sklearn.feature_extraction.text import CountVectorizer
# Join the stemmed words back into a single string for each message
d['s_joined'] = d['s'].apply(lambda x: ' '.join(x))


In [59]:
# Vectorize the text data
vectorizer = CountVectorizer()
X = vectorizer.fit_transform(d['s_joined'])
y = d['v1']


Splitting to train and test

In [56]:
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Apply naive bayes

In [57]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report


# Train a Naive Bayes classifier
nb_classifier = MultinomialNB()
nb_classifier.fit(X_train, y_train)

# Make predictions on the test set
y_pred = nb_classifier.predict(X_test)

# Evaluate the classifier
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

         ham       0.99      0.99      0.99       965
        spam       0.91      0.92      0.92       150

    accuracy                           0.98      1115
   macro avg       0.95      0.95      0.95      1115
weighted avg       0.98      0.98      0.98      1115



TfidfVectorizer

In [60]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Vectorize the text data using TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer()
X_tfidf = tfidf_vectorizer.fit_transform(d['s_joined'])

In [61]:
# Split the data into training and testing sets
X_train_tfidf, X_test_tfidf, y_train, y_test = train_test_split(X_tfidf, y, test_size=0.2, random_state=42)

In [62]:
# Train a Naive Bayes classifier
nb_classifier_tfidf = MultinomialNB()
nb_classifier_tfidf.fit(X_train_tfidf, y_train)

# Make predictions on the test set
y_pred_tfidf = nb_classifier_tfidf.predict(X_test_tfidf)

# Evaluate the classifier
print(classification_report(y_test, y_pred_tfidf))

              precision    recall  f1-score   support

         ham       0.96      1.00      0.98       965
        spam       1.00      0.71      0.83       150

    accuracy                           0.96      1115
   macro avg       0.98      0.86      0.91      1115
weighted avg       0.96      0.96      0.96      1115



# Conclusion
  Using count vectorizer we get a 98% accuracy  
  Using tfidf vectorizer we get a 96% accuracy